### PR TITLE
Added section for remote write key in secret and enabled monitoring

### PR DESCRIPTION
**Pull request by rohan47**
Added a new section for remote write key in managed fusion config secret and uncommented and changed the prometheus template as required.

### CODE REVIEW

**Comment by rohan47** [comment]
Unit tests are broken, will fix them after verifying the functionality of the code

**Comment by rohan47** [comment]
Tested the code and its functional, able to remote write to IBM Cloud Monitoring.
Need to fix the unit tests

**Review by ezio-auditore — COMMENTED** [review]
We might have to find a way to set the region of remote-write url depending on the rosa cluster region

**Comment by rohan47** [comment]
> We might have to find a way to set the region of remote-write url depending on the rosa cluster region

The endpoint will depend on the region of the monitoring instance. Will look into it more

**Comment by rohan47** [comment]
PR ready to be reviewed

File: templates/egressfirewall.go [hunk 1]
```diff
@@ -29,6 +29,12 @@ var EgressFirewallTemplate = ovnv1.EgressFirewall{
 				},
 				Type: ovnv1.EgressFirewallRuleAllow,
 			},
+			{
+				To: ovnv1.EgressFirewallDestination{
+					DNSName: "ingest.us-south.monitoring.cloud.ibm.com",
```
Comment (leelavg) [inline templates/egressfirewall.go:34]: - pls factor out this value in a const and use at other places
- if there's any mapping b/n AWS regions & IBM regions we could potentially set this value dynamically
- wherever the const will be defined, pls add a comment that it needs to be constructed dynamically at a later point

Comment (rohan47) [reply templates/egressfirewall.go]: Added a constant,
will look into dynamic construction as a future TODO and added a comment for the same

File: controllers/managedfusion_controller.go [hunk 2]
```diff
@@ -231,6 +236,12 @@ func (r *ManagedFusionReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 		return ctrl.Result{}, fmt.Errorf("%s.smtp_config is not in a valid yaml format: %v", managedFusionSecretName, err)
 	}
 
+	if config, found := r.managedFusionSecret.Data["remote_write_config"]; !found {
+		log.Info("Remote write config not present in the managed fusion secret")
```
[the remark this answers is not in the document]
Comment (leelavg) [reply controllers/managedfusion_controller.go:240]: - I still don't see removing remoteWrite stanza from prom if this key doesn't exist here
- I believe the effect will be at two places and we better log at those respective places:
1. checking for key existence (here itself)
2. key removed in a config update (possible at prom reconcile)

[the remark this answers is not in the document]
Comment (rohan47) [reply controllers/managedfusion_controller.go:240]: If the key is removed, Remote Write will stop sending metrics to IBM cloud monitoring as it won't authenticate.

File: controllers/managedfusion_controller.go [hunk 3]
```diff
@@ -432,6 +444,7 @@ func (r *ManagedFusionReconciler) reconcilePrometheus() error {
 
 		r.prometheus.Spec = desired.Spec
 		r.prometheus.Spec.ExternalLabels["clusterId"] = string(clusterVersion.Spec.ClusterID)
+		r.prometheus.Spec.RemoteWrite[0].BearerToken = r.remoteWriteConfig.BearerToken
```
Comment (rohan47) [inline controllers/managedfusion_controller.go:447]: If there is no key, this will simply be set to `""` and remote write will stop working, Agent won't complain about it
It will be logged in the code if the key is not present in the validation section

**Comment by rohan47** [comment]
ocs metrics exporter is not able to collect metrics
`oc logs -f ocs-metrics-exporter-6c4cb477cb-6lmcl`


```
W0419 13:20:56.123741       1 reflector.go:324] /remote-source/app/metrics/internal/collectors/cluster-advance-feature-use.go:166: failed to list *v1.CephObjectStore: cephobjectstores.ceph.rook.io is forbidden: User "system:serviceaccount:fusion-storage:ocs-metrics-exporter" cannot list resource "cephobjectstores" in API group "ceph.rook.io" at the cluster scope
E0419 13:20:56.123767       1 reflector.go:138] /remote-source/app/metrics/internal/collectors/cluster-advance-feature-use.go:166: Failed to watch *v1.CephObjectStore: failed to list *v1.CephObjectStore: cephobjectstores.ceph.rook.io is forbidden: User "system:serviceaccount:fusion-storage:ocs-metrics-exporter" cannot list resource "cephobjectstores" in API group "ceph.rook.io" at the cluster scope
W0419 13:20:57.398761       1 reflector.go:324] /remote-source/app/metrics/internal/collectors/ceph-cluster.go:54: failed to list *v1.CephCluster: cephclusters.ceph.rook.io is forbidden: User "system:serviceaccount:fusion-storage:ocs-metrics-exporter" cannot list resource "cephclusters" in API group "ceph.rook.io" at the cluster scope
E0419 13:20:57.398788       1 reflector.go:138] /remote-source/app/metrics/internal/collectors/ceph-cluster.go:54: Failed to watch *v1.CephCluster: failed to list *v1.CephCluster: cephclusters.ceph.rook.io is forbidden: User "system:serviceaccount:fusion-storage:ocs-metrics-exporter" cannot list resource "cephclusters" in API group "ceph.rook.io" at the cluster scope
W0419 13:20:58.502655       1 reflector.go:324] /remote-source/app/metrics/internal/collectors/cluster-advance-feature-use.go:166: failed to list *v1.CephRBDMirror: cephrbdmirrors.ceph.rook.io is forbidden: User "system:serviceaccount:fusion-storage:ocs-metrics-exporter" cannot list resource "cephrbdmirrors" in API group "ceph.rook.io" at the cluster scope
E0419 13:20:58.502682       1 reflector.go:138] /remote-source/app/metrics/internal/collectors/cluster-advance-feature-use.go:166: Failed to watch *v1.CephRBDMirror: failed to list *v1.CephRBDMirror: cephrbdmirrors.ceph.rook.io is forbidden: User "system:serviceaccount:fusion-storage:ocs-metrics-exporter" cannot list resource "cephrbdmirrors" in API group "ceph.rook.io" at the cluster scope
W0419 13:21:03.394832       1 reflector.go:324] /remote-source/app/metrics/internal/collectors/cluster-advance-feature-use.go:166: failed to list *v1.StorageClass: forbidden: User "system:serviceaccount:fusion-storage:ocs-metrics-exporter" cannot get path "/storageclasses"
E0419 13:21:03.394858       1 reflector.go:138] /remote-source/app/metrics/internal/collectors/cluster-advance-feature-use.go:166: Failed to watch *v1.StorageClass: failed to list *v1.StorageClass: forbidden: User "system:serviceaccount:fusion-storage:ocs-metrics-exporter" cannot get path "/storageclasses"
W0419 13:21:05.353238       1 reflector.go:324] /remote-source/app/metrics/internal/collectors/object-bucket.go:101: failed to list *v1.CephObjectStore: cephobjectstores.ceph.rook.io is forbidden: User "system:serviceaccount:fusion-storage:ocs-metrics-exporter" cannot list resource "cephobjectstores" in API group "ceph.rook.io" at the cluster scope
E0419 13:21:05.353270       1 reflector.go:138] /remote-source/app/metrics/internal/collectors/object-bucket.go:101: Failed to watch
```

**Comment by rohan47** [comment]
With the current changes the `prometheus-managed-fusion-prometheus-0` pod will throw the following error without the config but won't hamper the functionality of the agent

```
en: {\"errors\":[{\"reason\":\"Invalid credentials\",\"message\":\"Invalid credentials\"}]}"
ts=2023-04-19T18:27:18.331Z caller=dedupe.go:112 component=remote level=error remote_name=57d32b url=https://ingest.us-south.monitoring.cloud.ibm.com/prometheus/remote/write msg="non-recoverable error" count=1 exemplarCount=0 err="server returned HTTP status 403 Forbidden: {\"errors\":[{\"reason\":\"Invalid credentials\",\"message\":\"Invalid credentials\"}]}"
ts=2023-04-19T18:28:18.420Z caller=dedupe.go:112 component=remote level=error remote_name=57d32b url=https://ingest.us-south.monitoring.cloud.ibm.com/prometheus/remote/write msg="non-recoverable error" count=1 exemplarCount=0 err="server returned HTTP status 403 Forbidden: {\"errors\":[{\"reason\":\"Invalid credentials\",\"message\":\"Invalid credentials\"}]}"
```

Tried removing the remote write by following
```golang
if r.remoteWriteConfig.BearerToken != "" || r.prometheus.Spec.RemoteWrite != nil {
    r.prometheus.Spec.RemoteWrite[0].BearerToken = r.remoteWriteConfig.BearerToken
} else {
    r.prometheus.Spec.RemoteWrite = nil
}
```
Even removing the RemoteWrite doesn't work as we will have to delete the existing `prometheus-managed-fusion-prometheus-0` pod

**Review by leelavg — DISMISSED** [review]

**Comment by rohan47** [comment]
Need to disable remote write to remove the errors on the prometheus pod

File: controllers/managedfusion_controller.go [hunk 4]
```diff
@@ -432,6 +444,11 @@ func (r *ManagedFusionReconciler) reconcilePrometheus() error {
 
 		r.prometheus.Spec = desired.Spec
 		r.prometheus.Spec.ExternalLabels["clusterId"] = string(clusterVersion.Spec.ClusterID)
+		if r.remoteWriteConfig.BearerToken != "" {
+			r.prometheus.Spec.RemoteWrite[0].BearerToken = r.remoteWriteConfig.BearerToken
+		} else {
+			r.prometheus.Spec.RemoteWrite[0] = promv1.RemoteWriteSpec{}
```
Comment (rohan47) [inline controllers/managedfusion_controller.go:450]: With this change, remote write gets disabled and there is no error in the Prometheus pod

**Review by leelavg — APPROVED** [review]

File: controllers/managedfusion_controller.go [hunk 5]
```diff
@@ -231,6 +236,12 @@ func (r *ManagedFusionReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 		return ctrl.Result{}, fmt.Errorf("%s.smtp_config is not in a valid yaml format: %v", managedFusionSecretName, err)
 	}
 
+	if config, found := r.managedFusionSecret.Data["remote_write_config"]; !found {
+		log.Info("Remote write config not present in the managed fusion secret")
+	} else if err := yaml.Unmarshal(config, r.remoteWriteConfig); err != nil {
+		return ctrl.Result{}, fmt.Errorf("%s.remote_write_config is not in a valid yaml format: %v", managedFusionSecretName, err)
+	}
```
Comment (nb-ohad) [inline controllers/managedfusion_controller.go:243]: As this is an optional section, why do we need to log a message in the log on each reconciliation?
```suggestion
	if config, found := r.managedFusionSecret.Data["remote_write_config"]; found {
		r.remoteWriteConfig = &remoteWriteConfig{}
		if err := yaml.Unmarshal(config, r.remoteWriteConfig); err != nil {
			return ctrl.Result{}, fmt.Errorf("%s.remote_write_config is not in a valid yaml format: %v", managedFusionSecretName, err)
		}
	}
```

File: templates/prometheus.go [hunk 6]
```diff
@@ -34,42 +35,48 @@ var resourceSelector = metav1.LabelSelector{
...
+	"job:ceph_pools_iops_bytes:total",
+	"job:ceph_pools_iops:total",
+	"job:kube_pv:count",
+	"job:ceph_osd_metadata:count",
+	"ceph_health_status",
+	"ceph_cluster_total_used_raw_bytes",
+	"ceph_cluster_total_bytes",
+	"cluster:kubelet_volume_stats_used_bytes:provisioner:sum",
+	"cluster:kube_persistentvolumeclaim_resource_requests_storage_bytes:provisioner:sum",
+}
```
Comment (bindrad) [inline templates/prometheus.go:61]: These metrics are specific to DF and DFC offering so they should come from datafoundation or datafoundationclient package

File: controllers/managedfusion_controller.go [hunk 7]
```diff
@@ -294,6 +305,7 @@ func (r *ManagedFusionReconciler) initReconciler(ctx context.Context, req ctrl.R
 
 	r.smtpConfigData = &smtpConfig{}
 	r.pagerDutyConfigData = &pagerDutyConfig{}
+	r.remoteWriteConfig = &remoteWriteConfig{}
```
Comment (nb-ohad) [inline controllers/managedfusion_controller.go:308]: This is an optional section. As such we should not create an instance for the struct until we see a need for it. 
This will make sure that if the section is not present the value of `r.remoteWriteConfig` will stay `nil`

**Review by bindrad — CHANGES_REQUESTED** [review]

File: controllers/managedfusion_controller.go [hunk 8]
```diff
@@ -432,6 +444,11 @@ func (r *ManagedFusionReconciler) reconcilePrometheus() error {
 
 		r.prometheus.Spec = desired.Spec
 		r.prometheus.Spec.ExternalLabels["clusterId"] = string(clusterVersion.Spec.ClusterID)
+		if r.remoteWriteConfig.BearerToken != "" {
+			r.prometheus.Spec.RemoteWrite[0].BearerToken = r.remoteWriteConfig.BearerToken
+		} else {
+			r.prometheus.Spec.RemoteWrite[0] = promv1.RemoteWriteSpec{}
+		}
```
Comment (nb-ohad) [inline controllers/managedfusion_controller.go:451]: I believe the check here is wrong. 
This will disable remote write even in the case where there is a remote write config in the secret but the bearer token is missing. This will be considered a bug. 

You need to check for the existence of the section to decide if to enable or disable remote write. And as I have suggested in the couple of my previous comments that can be done by:
```suggestion
		if r.remoteWriteConfig != nil {
			if .r.remoteWriteConfig.BearerToken == "" {
				return fmt.Errorf("Invalid bearer token at remote write config")
			}
			r.prometheus.Spec.RemoteWrite[0].BearerToken = r.remoteWriteConfig.BearerToken
		} else {
			r.prometheus.Spec.RemoteWrite[0] = promv1.RemoteWriteSpec{}
		}
```

File: templates/prometheus.go [hunk 9]
```diff
@@ -34,42 +35,48 @@ var resourceSelector = metav1.LabelSelector{
...
+const (
+	//TODO: Construct monitoring dns name dynamically based on AWS and IBM Cloud Monitoring regions
+	MonitoringDNSName = "ingest.us-south.monitoring.cloud.ibm.com"
+)
+
 var (
 	KubeRBACProxyPortNumber             int    = 9339
 	PrometheusServingCertSecretName     string = "prometheus-serving-cert-secret"
 	PrometheusKubeRBACPoxyConfigMapName string = "prometheus-kube-rbac-proxy-config"
+	MonitoringURL                       string = fmt.Sprintf("https://%s/prometheus/remote/write", MonitoringDNSName)
```
Comment (nb-ohad) [inline templates/prometheus.go:47]: Why do we need to dynamically construct a string here where the entire endpoint (not URL) is static?
Also, this is a remote wWrite endpoint, not a monitoring endpoint.
```suggestion
	remoteWriteEndpoint                     string = fmt.Sprintf("https://ingest.us-south.monitoring.cloud.ibm.com/prometheus/remote/write", MonitoringDNSName)
```

File: templates/prometheus.go (hunk 6 above)
Comment (nb-ohad) [inline templates/prometheus.go:61]: Why these metrics? where did the requirement for these come from (in the context of the new agent)?
Also, I think the most important metric is missing.

File: templates/prometheus.go (hunk 9 above)
Comment (leelavg) [reply templates/prometheus.go:47]: This endpoint was being used in a couple of places and so I have asked to use a const.

Comment (rohan47) [reply templates/prometheus.go:47]: Will make this static for now

File: templates/prometheus.go (hunk 6 above)
Comment (rohan47) [reply templates/prometheus.go:61]: These are the metrics that were exposed to RHOBS in the deployer

Comment (rohan47) [reply templates/prometheus.go:61]: @nb-ohad I can think of two metrics right now one is `agent_installation_status` and `offering_installation_status`, what other metrics are expected to be exposed?

File: controllers/managedfusion_controller.go [hunk 10]
```diff
@@ -98,6 +98,7 @@ type ManagedFusionReconciler struct {
 	pagerDutyConfigData          *pagerDutyConfig
 	egressNetworkPolicy          *openshiftv1.EgressNetworkPolicy
 	egressFirewall               *ovnv1.EgressFirewall
+	remoteWriteConfig            *remoteWriteConfig
```
[the remark this answers is not in the document]
Comment (rohan47) [reply controllers/managedfusion_controller.go]: Done

File: templates/prometheus.go (hunk 6 above)
Comment (rohan47) [reply templates/prometheus.go]: Moved ceph related metrics and alerts to `datafoundation/templates/storagecluster.go`

File: templates/prometheus.go [hunk 11]
```diff
@@ -34,42 +35,48 @@ var resourceSelector = metav1.LabelSelector{
...
+	"CephOSDVersionMismatch",
+	"CephMonVersionMismatch",
+	"CephOSDFlapping",
+	"CephOSDDiskNotResponding",
+	"CephOSDDiskUnavailable",
+	"CephDataRecoveryTakingTooLong",
+	"CephPGRepairTakingTooLong",
+	"CephMonQuorumAtRisk",
+	"CephMonQuorumLost",
+}
```
[the remark this answers is not in the document]
Comment (rohan47) [reply templates/prometheus.go]: moved this to `datafoundation/templates/storagecluster.go`

File: controllers/managedfusion_controller.go (hunk 8 above)
Comment (rohan47) [reply controllers/managedfusion_controller.go]: Made the change as per suggestion

File: templates/prometheus.go (hunk 6 above)
Comment (nb-ohad) [reply templates/prometheus.go]: Why do we need these right now? we are missing a lot of infrastructure to support these. I will keep the entire list in a comment for this PR

Comment (rohan47) [reply templates/prometheus.go]: Okay, So what metrics do we need to expose from the agent side for now?

File: controllers/managedfusion_controller.go [hunk 12]
```diff
@@ -432,6 +444,16 @@ func (r *ManagedFusionReconciler) reconcilePrometheus() error {
 
 		r.prometheus.Spec = desired.Spec
 		r.prometheus.Spec.ExternalLabels["clusterId"] = string(clusterVersion.Spec.ClusterID)
+
+		if r.remoteWriteConfigData != nil {
+			if r.remoteWriteConfigData.BearerToken == "" {
+				return fmt.Errorf("bearer token for remote write config data is invalid")
```
Comment (nb-ohad) [inline controllers/managedfusion_controller.go:450]: Let's keep invalid field error similar
```suggestion
				return fmt.Errorf("invalid remote write bearer token, empty string")
```

File: datafoundation/templates/storagecluster.go [hunk 13]
```diff
@@ -31,6 +31,34 @@ const (
...
+	// "CephOSDVersionMismatch",
+	// "CephMonVersionMismatch",
+	// "CephOSDFlapping",
+	// "CephOSDDiskNotResponding",
+	// "CephOSDDiskUnavailable",
+	// "CephDataRecoveryTakingTooLong",
+	// "CephPGRepairTakingTooLong",
+	// "CephMonQuorumAtRisk",
+	// "CephMonQuorumLost",
+}
```
Comment (nb-ohad) [inline datafoundation/templates/storagecluster.go:60]: Why are these here? Let's keep everything where it was until we have a mechanism to allow different offering to provide their own metrics dynamically (outside the scope of this PR). I don't like the fact, that with this change, generalized code needs to statically import the storagecluster template.  Not to mention that this has nothing to do with the storagecluster itself.

File: templates/egressnetworkpolicy.go
Comment (nb-ohad) [inline templates/egressnetworkpolicy.go:1]: This entire file is not used anymore and should have been removed in a previous commit. Let's not add anything else here, and if we can let's have a PR to remove it.

File: templates/prometheus.go [hunk 14]
```diff
@@ -34,42 +36,25 @@ var resourceSelector = metav1.LabelSelector{
...
+	//TODO: Construct monitoring dns name dynamically based on AWS and IBM Cloud Monitoring regions
+	RemoteWriteDNSName = "ingest.us-south.monitoring.cloud.ibm.com"
+)
+
 var (
 	KubeRBACProxyPortNumber             int    = 9339
 	PrometheusServingCertSecretName     string = "prometheus-serving-cert-secret"
 	PrometheusKubeRBACPoxyConfigMapName string = "prometheus-kube-rbac-proxy-config"
+	RemoteWriteUrl                      string = "https://ingest.us-south.monitoring.cloud.ibm.com/prometheus/remote/write"
 )
```
Comment (nb-ohad) [inline templates/prometheus.go:49]: All of these are consts, why are we using vars?

File: templates/prometheus.go [hunk 15]
```diff
@@ -34,42 +36,25 @@ var resourceSelector = metav1.LabelSelector{
...
+const (
+	//TODO: Construct monitoring dns name dynamically based on AWS and IBM Cloud Monitoring regions
+	RemoteWriteDNSName = "ingest.us-south.monitoring.cloud.ibm.com"
+)
+
 var (
 	KubeRBACProxyPortNumber             int    = 9339
 	PrometheusServingCertSecretName     string = "prometheus-serving-cert-secret"
 	PrometheusKubeRBACPoxyConfigMapName string = "prometheus-kube-rbac-proxy-config"
+	RemoteWriteUrl                      string = "https://ingest.us-south.monitoring.cloud.ibm.com/prometheus/remote/write"
```
Comment (nb-ohad) [inline templates/prometheus.go:48]: This one is only used within this file and only at one place, please inline it.

**Review by nb-ohad — CHANGES_REQUESTED** [review]

File: templates/prometheus.go (hunk 6 above)
Comment (nb-ohad) [reply templates/prometheus.go]: none. all work on specific metrics to come later on based on actual use cases

File: controllers/managedfusion_controller.go [hunk 16]
```diff
@@ -231,6 +236,13 @@ func (r *ManagedFusionReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 		return ctrl.Result{}, fmt.Errorf("%s.smtp_config is not in a valid yaml format: %v", managedFusionSecretName, err)
 	}
 
+	if config, found := r.managedFusionSecret.Data["remote_write_config"]; found {
+		r.remoteWriteConfigData = &remoteWriteConfigData{}
+		if err := yaml.Unmarshal(config, r.remoteWriteConfigData); err != nil {
+			return ctrl.Result{}, fmt.Errorf("%s.remote_write_config is not in a valid yaml format: %v", managedFusionSecretName, err)
```
Comment (nb-ohad) [inline controllers/managedfusion_controller.go:242]: Errors are not log messages and should be structured accordingly. And we should not assume that the payload is invalid just because unmarshal didn't work.
```suggestion
			return ctrl.Result{}, fmt.Errorf("Failed to unmarshal agent remote_write_config at %s,  %v", managedFusionSecretName, err)
```

And please apply the same for the equivalent errors above this one.

**Review by nb-ohad — CHANGES_REQUESTED** [review]

**Review by bindrad — CHANGES_REQUESTED** [review]

File: controllers/managedfusion_controller.go (hunk 16 above)
Comment (rohan47) [reply controllers/managedfusion_controller.go]: Done

File: controllers/managedfusion_controller.go [hunk 17]
```diff
@@ -113,6 +114,10 @@ type pagerDutyConfig struct {
 	ServiceKey  string `yaml:"serviceKey"`
 }
 
+type remoteWriteConfigData struct {
```
[the remark this answers is not in the document]
Comment (rohan47) [reply controllers/managedfusion_controller.go]: Done

**Review by bindrad — DISMISSED** [review]

File: controllers/managedfusion_controller.go (hunk 16 above)
Comment (nb-ohad) [reply controllers/managedfusion_controller.go]: @rohan47 I see only a change for the message you added. The ask above was:
```And please apply the same for the equivalent errors above this one.```

Please fix
